### ISSUE TITLE
Duplicate GlobalKey error when updating child during layout

### ISSUE THREAD

**Issue by AkramIzz** [flutter/flutter#61248]
Hi! I'm trying to write a layout builder widget that passes another widget's size to build its child. It kinda works similarly to LayoutBuilder, except it takes a child argument which is the widget that its size is passed to the builder function. To measure the child's size, it's added as a child to the layout builder widget using updateChild method, then at the layout phase, after laying out the child, its size (and the widget itself) is passed to the builder (indirectly using invokeLayoutCallback and owner.buildScope). Finally, the widget returned by the builder is used to update the child using updateChild.

It works as expected but if the child widget is used in the builder it will result in inflating and laying out the widget twice. To prevent this I used a wrapping widget for the child widget with a global key to allow the framework to do its thing and move the element to its new place in the tree. [The end result](https://gist.github.com/AkramIzz/265ca4e618f731f8e8d3bec7406ac462). 

Finally to the error.

## Steps to Reproduce

1. Run `flutter create bug`.
2. include the file in lib provided above. 
3. change main.dart to the following
```dart
import 'package:flutter/material.dart';
import 'size_layout_builder.dart';

final firstKey = GlobalKey(debugLabel: 'first');

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(TestApp());
}

class TestApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: TestWidget(),
      ),
    );
  }
}

class TestWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SizeLayoutBuilder(
      child: Container(
        key: firstKey,
        padding: const EdgeInsets.all(32.0),
        color: Colors.green,
      ),
      builder: (context, size, _) {
        return Container(
          padding: const EdgeInsets.all(32.0),
          color: Colors.blue,
          child: Container(
            key: firstKey,
            padding: const EdgeInsets.all(32.0),
            color: Colors.green,
          ),
        );
      },
    );
  }
}

```

**Expected results:**
Expect no errors, correct me if I'm wrong, but since the global key of child widget and widget with same key in builder don't exist in the tree at the same time (once the tree is finalized) there should not be any errors

**Actual results:**
An exception is thrown by the framework stating that Multiple widgets used the same GlobalKey (firstKey). If I removed the firstKey global key, *or the wrapping widget's key* the error disappears.
```
════════ Exception caught by widgets library ═══════════════════════════════════
The following assertion was thrown while finalizing the widget tree:
Multiple widgets used the same GlobalKey.

The key [GlobalKey#56c45 first] was used by multiple widgets. The parents of those widgets were:
- _WrapperWidget-[GlobalKey#76b2a SizeLayoutBuilderChildWrapperKey]
- Padding(padding: EdgeInsets.all(32.0), dependencies: [Directionality], renderObject: RenderPadding#f0c9c relayoutBoundary=up4)
A GlobalKey can only be specified on one widget at a time in the widget tree.
When the exception was thrown, this was the stack
#0      GlobalKey._debugVerifyGlobalKeyReservation.<anonymous closure>.<anonymous closure>.<anonymous closure> 
package:flutter/…/widgets/framework.dart:246
#1      _LinkedHashMapMixin.forEach  (dart:collection-patch/compact_hash.dart:379:8)
#2      GlobalKey._debugVerifyGlobalKeyReservation.<anonymous closure>.<anonymous closure> 
package:flutter/…/widgets/framework.dart:193
#3      _LinkedHashMapMixin.forEach  (dart:collection-patch/compact_hash.dart:379:8)
#4      GlobalKey._debugVerifyGlobalKeyReservation.<anonymous closure> 
package:flutter/…/widgets/framework.dart:189
...
════════════════════════════════════════════════════════════════════════════════

════════ Exception caught by widgets library ═══════════════════════════════════
Multiple widgets used the same GlobalKey.
════════════════════════════════════════════════════════════════════════════════```
```

<details>
  <summary>Logs</summary>

```
~$ flutter doctor -v

[✓] Flutter (Channel stable, v1.17.5, on Linux, locale en_US.UTF-8)
    • Flutter version 1.17.5 at /home/akram/Desktop/tools/flutter/flutter
    • Framework revision 8af6b2f038 (10 days ago), 2020-06-30 12:53:55 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /home/akram/Desktop/tools/android/Sdk
    • Platform android-29, build-tools 28.0.3
    • ANDROID_HOME = /home/akram/Desktop/tools/android/Sdk
    • Java binary at: /home/akram/Desktop/tools/android/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] Android Studio (version 3.3)
    • Android Studio at /home/akram/Desktop/tools/android/android-studio
    • Flutter plugin version 23.2.1
    • Dart plugin version 171.4424
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.39.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.7.1

[✓] Connected device (1 available)
    • AMN LX9 • 7YBNU19816107646 • android-arm • Android 9 (API 28)

• No issues found!
```

</details>

**Comment by TahaTesser** [comment]
<details> 
<summary>Code Sample</summary> 

``` 
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

final firstKey = GlobalKey(debugLabel: 'first');

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(TestApp());
}

class TestApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: TestWidget(),
      ),
    );
  }
}

class TestWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SizeLayoutBuilder(
      child: Container(
        key: firstKey,
        padding: const EdgeInsets.all(32.0),
        color: Colors.green,
      ),
      builder: (context, size, _) {
        return Container(
          padding: const EdgeInsets.all(32.0),
          color: Colors.blue,
          child: Container(
            key: firstKey,
            padding: const EdgeInsets.all(32.0),
            color: Colors.green,
          ),
        );
      },
    );
  }
}

class _WrapperWidget extends StatelessWidget {
  _WrapperWidget({this.child, Key key}) : super(key: key);

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return child;
  }
}

/// The signature of the [SizeLayoutBuilder] builder function.
typedef LayoutWidgetBuilder = Widget Function(
    BuildContext context, Size size, Widget child);

/// An abstract superclass for widgets that defer their building until layout.
///
/// Similar to the [Builder] widget except that the framework calls the [builder]
/// function at layout time and provides the size of the child widget.
class SizeLayoutBuilder extends RenderObjectWidget {
  /// Creates a widget that defers its building until layout.
  ///
  /// The [builder] and [child] arguments must not be null,
  /// and the returned widget should not be null.
  const SizeLayoutBuilder({
    Key key,
    @required this.child,
    @required this.builder,
  })  : assert(builder != null),
        assert(child != null),
        super(key: key);

  @override
  _SizeLayoutBuilderElement createElement() => _SizeLayoutBuilderElement(this);

  @override
  _RenderSizeLayoutBuilder createRenderObject(BuildContext context) =>
      _RenderSizeLayoutBuilder();

  /// Called at layout time to construct the widget tree.
  ///
  /// The builder must not return null.
  final LayoutWidgetBuilder builder;

  final Widget child;

  // updateRenderObject is redundant with the logic in the LayoutBuilderElement below.
}

class _SizeLayoutBuilderElement extends RenderObjectElement {
  _SizeLayoutBuilderElement(SizeLayoutBuilder widget) : super(widget);

  @override
  SizeLayoutBuilder get widget => super.widget;

  @override
  _RenderSizeLayoutBuilder get renderObject => super.renderObject;

  Element _child;

  // used to wrap widget.child with a global key so that moving it in widget.builder
  // doesn't result in rebuilding the child.
  final childKey = GlobalKey(debugLabel: 'SizeLayoutBuilderChildWrapperKey');

  @override
  void visitChildren(ElementVisitor visitor) {
    if (_child != null) visitor(_child);
  }

  @override
  void forgetChild(Element child) {
    super.forgetChild(child);
    assert(child == _child);
    _child = null;
  }

  @override
  void mount(Element parent, dynamic newSlot) {
    super.mount(parent, newSlot); // Creates the renderObject.
    onWidgetChildChanged();
    renderObject.updateCallback(_layout);
  }

  @override
  void update(SizeLayoutBuilder newWidget) {
    assert(widget != newWidget);
    super.update(newWidget);
    assert(widget == newWidget);
    onWidgetChildChanged();
    renderObject.updateCallback(_layout);
    renderObject.markNeedsLayout();
  }

  void onWidgetChildChanged() {
    // optimization for when this widget is potentially moved in widget.builder
    // since flutter doesn't have tree diffing moving the child without using a global widget
    // will result in the widget rebuilding itself unncessarily.
    final wrapper = _WrapperWidget(key: childKey, child: widget.child);
    _child = updateChild(_child, wrapper, null);
  }

  @override
  void performRebuild() {
    // This gets called if markNeedsBuild() is called on us.
    // That might happen if, e.g., our builder uses Inherited widgets.
    renderObject.markNeedsLayout();
    super
        .performRebuild(); // Calls widget.updateRenderObject (a no-op in this case).
  }

  @override
  void unmount() {
    renderObject.updateCallback(null);
    super.unmount();
  }

  void _layout(Size size) {
    owner.buildScope(this, () {
      Widget built;
      if (widget.builder != null) {
        try {
          built = widget.builder(this, size, _child.widget);
          debugWidgetBuilderValue(widget, built);
        } catch (e, stack) {
          built = ErrorWidget.builder(
            _debugReportException(
              ErrorDescription('building $widget'),
              e,
              stack,
              informationCollector: () sync* {
                yield DiagnosticsDebugCreator(DebugCreator(this));
              },
            ),
          );
        }
      }
      try {
        _child = updateChild(_child, built, null);
        assert(_child != null);
      } catch (e, stack) {
        built = ErrorWidget.builder(
          _debugReportException(
            ErrorDescription('building $widget'),
            e,
            stack,
            informationCollector: () sync* {
              yield DiagnosticsDebugCreator(DebugCreator(this));
            },
          ),
        );
        _child = updateChild(null, built, slot);
      }
    });
  }

  @override
  void insertChildRenderObject(RenderObject child, dynamic slot) {
    final RenderObjectWithChildMixin<RenderObject> renderObject =
        this.renderObject;
    assert(slot == null);
    assert(renderObject.debugValidateChild(child));
    renderObject.child = child;
    assert(renderObject == this.renderObject);
  }

  @override
  void moveChildRenderObject(RenderObject child, dynamic slot) {
    assert(false);
  }

  @override
  void removeChildRenderObject(RenderObject child) {
    final _RenderSizeLayoutBuilder renderObject = this.renderObject;
    assert(renderObject.child == child);
    renderObject.child = null;
    assert(renderObject == this.renderObject);
  }
}

class _RenderSizeLayoutBuilder extends RenderBox
    with RenderObjectWithChildMixin<RenderBox> {
  void Function(Size size) _callback;

  /// Change the layout callback.
  void updateCallback(void Function(Size size) value) {
    if (value == _callback) return;
    _callback = value;
    markNeedsLayout();
  }

  /// Invoke the layout callback.
  void layoutAndBuildChild(Size size) {
    assert(_callback != null);

    invokeLayoutCallback((constraints) {
      _callback(size);
    });
  }

  @override
  double computeMinIntrinsicWidth(double height) {
    assert(_debugThrowIfNotCheckingIntrinsics());
    return 0.0;
  }

  @override
  double computeMaxIntrinsicWidth(double height) {
    assert(_debugThrowIfNotCheckingIntrinsics());
    return 0.0;
  }

  @override
  double computeMinIntrinsicHeight(double width) {
    assert(_debugThrowIfNotCheckingIntrinsics());
    return 0.0;
  }

  @override
  double computeMaxIntrinsicHeight(double width) {
    assert(_debugThrowIfNotCheckingIntrinsics());
    return 0.0;
  }

  @override
  void performLayout() {
    assert(child != null);
    child.layout(constraints, parentUsesSize: true);
    layoutAndBuildChild(child.size);
    child.layout(constraints, parentUsesSize: true);
    size = constraints.constrain(child.size);
  }

  @override
  bool hitTestChildren(BoxHitTestResult result, {Offset position}) {
    return child?.hitTest(result, position: position) ?? false;
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    assert(child != null);
    context.paintChild(child, offset);
  }

  bool _debugThrowIfNotCheckingIntrinsics() {
    assert(() {
      if (!RenderObject.debugCheckingIntrinsics) {
        throw FlutterError(
            'SizeLayoutBuilder does not support returning intrinsic dimensions.\n'
            'Calculating the intrinsic dimensions would require running the layout '
            'callback speculatively, which might mutate the live render object tree.');
      }
      return true;
    }());

    return true;
  }
}

FlutterErrorDetails _debugReportException(
  DiagnosticsNode context,
  dynamic exception,
  StackTrace stack, {
  InformationCollector informationCollector,
}) {
  final FlutterErrorDetails details = FlutterErrorDetails(
    exception: exception,
    stack: stack,
    context: context,
    informationCollector: informationCollector,
  );
  FlutterError.reportError(details);
  return details;
}

``` 

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel dev, 1.20.0-7.1.pre, on Mac OS X 10.15.5 19F101, locale
    en-GB)
    • Flutter version 1.20.0-7.1.pre at /Users/taha/Code/flutter_dev
    • Framework revision 7736f3bc90 (2 days ago), 2020-07-10 16:33:05 -0700
    • Engine revision d48085141c
    • Dart version 2.9.0 (build 2.9.0-21.2.beta)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/taha/Code/sdk
    • Platform android-30, build-tools 30.0.0
    • ANDROID_HOME = /Users/taha/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (4 available)
    • SM M305F (mobile) • 32003c30dc19668f • android-arm64  • Android 10 (API
      29)
    • macOS (desktop)   • macos            • darwin-x64     • Mac OS X 10.15.5
      19F101
    • Web Server (web)  • web-server       • web-javascript • Flutter Tools
    • Chrome (web)      • chrome           • web-javascript • Google Chrome
      83.0.4103.116

• No issues found!
```

</details>

**Comment by goderbauer** [comment]
Can you try to reduce your code sample to isolate the error? SizeLayoutBuilder is ~300 lines, which are probably not all relevant.

**Comment by AkramIzz** [comment]
Well, I wanted to provide a working example. Most of the clutter is required by the RenderObject/Element protocol or comments to help in understanding the code. Here are the relevant parts just so you can read the gist of it, but it won't compile
```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

typedef LayoutWidgetBuilder = Widget Function(BuildContext context, Size size, Widget child);

class SizeLayoutBuilder extends RenderObjectWidget {
  const SizeLayoutBuilder({
    Key key,
    @required this.child,
    @required this.builder,
  }) : super(key: key);

  final LayoutWidgetBuilder builder;

  final Widget child;
}

class _SizeLayoutBuilderElement extends RenderObjectElement {
  _SizeLayoutBuilderElement(SizeLayoutBuilder widget) : super(widget);

  @override
  SizeLayoutBuilder get widget => super.widget;

  @override
  _RenderSizeLayoutBuilder get renderObject => super.renderObject;

  Element _child;

  // used to wrap widget.child with a global key so that moving it in widget.builder
  // doesn't result in rebuilding the child.
  final childKey = GlobalKey(debugLabel: 'SizeLayoutBuilderChildWrapperKey');

  @override
  void mount(Element parent, dynamic newSlot) {
    super.mount(parent, newSlot); // Creates the renderObject.
    onWidgetChildChanged();
    renderObject.updateCallback(_layout);
  }

  @override
  void update(SizeLayoutBuilder newWidget) {
    assert(widget != newWidget);
    super.update(newWidget);
    assert(widget == newWidget);
    onWidgetChildChanged();
    renderObject.updateCallback(_layout);
    renderObject.markNeedsLayout();
  }

  void onWidgetChildChanged() {
    final wrapper = _WrapperWidget(key: childKey, child: widget.child);
    _child = updateChild(_child, wrapper, null);
  }

  @override
  void performRebuild() {
    renderObject.markNeedsLayout();
    super.performRebuild();
  }

  void _layout(Size size) {
    owner.buildScope(this, () {
      Widget built;
      if (widget.builder != null) {
        built = widget.builder(this, size, _child.widget);
      }
      _child = updateChild(_child, built, null);
    });
  }
}

class _RenderSizeLayoutBuilder extends RenderBox
    with RenderObjectWithChildMixin<RenderBox> {
  void Function(Size size) _callback;

  void updateCallback(void Function(Size size) value) {
    if (value == _callback) return;
    _callback = value;
    markNeedsLayout();
  }

  void layoutAndBuildChild(Size size) {
    assert(_callback != null);

    invokeLayoutCallback((constraints) {
      _callback(size);
    });
  }

  @override
  void performLayout() {
    assert(child != null);
    child.layout(constraints, parentUsesSize: true);
    layoutAndBuildChild(child.size);
    child.layout(constraints, parentUsesSize: true);
    size = constraints.constrain(child.size);
  }
}
```

**Comment by xu-baolin** [comment]
@AkramIzz This issue has been fixed recently, via #62323

**Comment by TahaTesser** [comment]
Hi @AkramIzz 
i can confirm this issue  is fixed on latest `master`

<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel master, 1.21.0-6.0.pre.160, on Mac OS X 10.15.6 19G73,
    locale en-GB)
    • Flutter version 1.21.0-6.0.pre.160 at /Users/taha/Code/flutter_master
    • Framework revision 31ee51a302 (10 hours ago), 2020-08-03 17:27:21 -0700
    • Engine revision fa16c3d0ba
    • Dart version 2.10.0 (build 2.10.0-2.0.dev 0f0e04ec3a)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/taha/Code/sdk
    • Platform android-30, build-tools 30.0.1
    • ANDROID_HOME = /Users/taha/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.1

 
[✓] Connected device (5 available)            
    • Taha’s iPhone (mobile) • 00008020-001059882212002E            • ios            • iOS 13.6
    • iPhone 11 (mobile)     • 1EB615AF-5AB9-4AAA-A587-A6642BB79484 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-13-6 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64     • Mac OS X
      10.15.6 19G73
    • Web Server (web)       • web-server                           • web-javascript • Flutter Tools
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome
      84.0.4147.105

• No issues found!
```

</details> 

Closing, as fixed, please write in the comments and I will reopen it.
Thank you
